ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala

```diff
@@ -24,111 +21,166 @@ object MetricHandle {
+    def gauge[T](name: MetricName, initial: T): Gauge[T]
 
-    def addGauge[T <: codahale.Gauge[M], M](
+    def gaugeWithSupplier[T](
         name: MetricName,
-        newGauge: => T,
-        resetGauge: (T => Unit),
-    ): Gauge[T, M] = gauge(name, registry.register(name, newGauge), resetGauge)
+        gaugeSupplier: () => () => T,
+    ): Unit
 
-    def gaugeWithSupplier[T <: codahale.Gauge[M], M](
-        name: MetricName,
-        gaugeSupplier: MetricSupplier[codahale.Gauge[_]],
-    ): Gauge[T, M] =
-      gauge(name, registry.gauge(name, gaugeSupplier).asInstanceOf[T], _ => ())
+    def meter(name: MetricName): Meter
+
+    def counter(name: MetricName): Counter
+
+    def histogram(name: MetricName): Histogram
+
+  }
+  trait DropwizardFactory extends Factory {
+
+    def prefix: MetricName
 
-    def gauge[T <: codahale.Gauge[M], M](
+    def registry: codahale.MetricRegistry
+
+    def timer(name: MetricName): Timer = DropwizardTimer(name, registry.timer(name))
+
+    def gauge[T](name: MetricName, initial: T): Gauge[T] =
+      synchronized {
+        registry.remove(name)
+        val registeredGauge = registry.register(name, Gauges.VarGauge(initial))
+        DropwizardGauge(name, registeredGauge)
+      }
+
+    def gaugeWithSupplier[T](
         name: MetricName,
-        registerGauge: => T,
-        resetGauge: (T => Unit),
-    ): Gauge[T, M] = blocking {
+        gaugeSupplier: () => () => T,
+    ): Unit =
       synchronized {
         registry.remove(name)
-        val res: Gauge[T, M] = {
-          val gauge = registerGauge
-          Gauge(name, gauge)
-        }
-        resetGauge(res.metric)
-        res
+        val _ = registry.gauge(
+          name,
+          () => {
+            val valueGetter = gaugeSupplier()
+            new codahale.Gauge[T] { override def getValue: T = valueGetter() }
+          },
+        )
+        ()
       }
-    }
 
     def meter(name: MetricName): Meter = {
       // This is idempotent
-      Meter(name, registry.meter(name))
+      DropwizardMeter(name, registry.meter(name))
     }
 
     def counter(name: MetricName): Counter = {
       // This is idempotent
-      Counter(name, registry.counter(name))
+      DropwizardCounter(name, registry.counter(name))
     }
 
     def histogram(name: MetricName): Histogram = {
-      Histogram(name, registry.histogram(name))
+      DropwizardHistogram(name, registry.histogram(name))
     }
 
   }
 
-  trait FactoryWithDBMetrics extends MetricHandle.Factory {
+  trait FactoryWithDBMetrics extends MetricHandle.DropwizardFactory {
     def createDbMetrics(name: String): DatabaseMetrics =
       new DatabaseMetrics(prefix, name, registry)
   }
 
-  sealed case class Timer(name: String, metric: codahale.Timer)
-      extends MetricHandle[codahale.Timer] {
+  sealed trait Timer extends MetricHandle {
+
     def metricType: String = "Timer"
 
-    def timeEitherT[E, A](ev: EitherT[Future, E, A]): EitherT[Future, E, A] = {
```

Comment:
breaks canton if removed

Reply:
It should be moved together with themethods from the `TimedLoadGauge` as it seems to target the same API.